### PR TITLE
Feature/fga/voip timeline

### PR DESCRIPTION
**Pull request by ganfra**
This PR introduces new timeline tiles for call events. It will only show one tile for a call depending of its current state.
photo_2020-12-09 18 27 37

### CODE REVIEW

File: vector/src/main/java/im/vector/app/core/extensions/TextView.kt [hunk 1]
```diff
@@ -48,11 +56,13 @@ fun TextView.setTextOrHide(newText: CharSequence?, hideWhenBlank: Boolean = true
  * @param coloredTextRes the resource id of the colored part of the text
  * @param colorAttribute attribute of the color. Default to colorAccent
  * @param underline true to also underline the text. Default to false
+ * @param onClick attributes to handle click on the colored part if needed required
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/extensions/TextView.kt:59]: "needed required" ? :p

File: vector/src/main/java/im/vector/app/core/extensions/TextView.kt [hunk 2]
```diff
@@ -65,12 +75,39 @@ fun TextView.setTextWithColoredPart(@StringRes fullTextRes: Int,
     text = SpannableString(fullText)
             .apply {
                 setSpan(foregroundSpan, index, index + coloredPart.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+                if (onClick != null) {
+                    val clickableSpan = object : ClickableSpan() {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/extensions/TextView.kt:79]: Create a class `NoUnderlineClickableSpan` ?

File: vector/src/main/java/im/vector/app/core/extensions/TextView.kt [hunk 3]
```diff
@@ -65,12 +75,39 @@ fun TextView.setTextWithColoredPart(@StringRes fullTextRes: Int,
...
+                        override fun onClick(widget: View) {
+                            onClick()
+                        }
+
+                        override fun updateDrawState(ds: TextPaint) {
+                            ds.color = color
+                            // underline will be handled separately if needed, see below
+                            ds.isUnderlineText = false
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/core/extensions/TextView.kt:87]: Replace by `ds.isUnderlineText = !underline` to ensure that the clickable part is always underlined?

File: vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailFragment.kt [hunk 4]
```diff
@@ -368,6 +368,7 @@ class RoomDetailFragment @Inject constructor(
                 is RoomDetailViewEvents.DisplayEnableIntegrationsWarning -> displayDisabledIntegrationDialog()
                 is RoomDetailViewEvents.OpenIntegrationManager           -> openIntegrationManager()
                 is RoomDetailViewEvents.OpenFile                         -> startOpenFileIntent(it)
+                is RoomDetailViewEvents.DisplayAndAcceptCall ->             acceptIncomingCall(it)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailFragment.kt:371]: formatting...

File: vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt [hunk 5]
```diff
@@ -269,6 +269,7 @@ class RoomDetailViewModel @AssistedInject constructor(
             is RoomDetailAction.SelectStickerAttachment          -> handleSelectStickerAttachment()
             is RoomDetailAction.OpenIntegrationManager           -> handleOpenIntegrationManager()
             is RoomDetailAction.StartCall                        -> handleStartCall(action)
+            is RoomDetailAction.AcceptCall -> handleAcceptCall(action)
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/RoomDetailViewModel.kt:272]: format...

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/TimelineItemFactory.kt [hunk 6]
```diff
@@ -84,7 +87,7 @@ class TimelineItemFactory @Inject constructor(private val messageItemFactory: Me
                 EventType.KEY_VERIFICATION_KEY,
                 EventType.KEY_VERIFICATION_READY,
                 EventType.KEY_VERIFICATION_MAC,
-                EventType.CALL_CANDIDATES       -> {
+                EventType.CALL_CANDIDATES -> {
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/factory/TimelineItemFactory.kt:90]: please revert fomartting change

File: vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/CallTileTimelineItem.kt [hunk 7]
```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2019 New Vector Ltd
```
Comment (bmarty) [inline vector/src/main/java/im/vector/app/features/home/room/detail/timeline/item/CallTileTimelineItem.kt:2]: 2020

File: vector/src/main/res/layout/item_timeline_event_base_state.xml [hunk 8]
```diff
@@ -50,7 +56,7 @@
             android:id="@+id/messageVerificationDoneStub"
             style="@style/TimelineContentStubBaseParams"
             android:layout="@layout/item_timeline_event_status_tile_stub"
-            tools:visibility="visible" />
+            tools:visibility="gone" />
```
Comment (bmarty) [inline vector/src/main/res/layout/item_timeline_event_base_state.xml:59]: I prefer to keep the visibility for easier debugging/editting.
You could add `tools:layout_marginTop` attribute to ensure the views does not overlap. Also applicable to the other ViewStub.

File: vector/src/main/res/layout/item_timeline_event_call_tile_stub.xml [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:orientation="vertical">
+
+    <ImageView
+        android:id="@+id/itemCallCreatorAvatar"
+        android:layout_width="40dp"
+        android:layout_height="40dp"
```
Comment (bmarty) [inline vector/src/main/res/layout/item_timeline_event_call_tile_stub.xml:12]: Add tools avatar?

**Review by bmarty — CHANGES_REQUESTED** [review]
That's great improvement on the timeline, thanks!
Some minor remarks

File: vector/src/main/java/im/vector/app/core/extensions/TextView.kt (hunk 2 above)
Comment (ganfra) [reply vector/src/main/java/im/vector/app/core/extensions/TextView.kt:79]: Not sure it's relevant

**Pull request merged**